elyra/pipeline/pipeline_definition.py

```diff
@@ -0,0 +1,281 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import json
+import os
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+
+class InvalidPipeline(Exception):
+    pass
+
+
+class NodeBase():
+    _pipeline_definition: Dict = None
+    _node: Dict = None
+
+    def __init__(self, pipeline_definition: Dict, node: Dict):
+        self._pipeline_definition = pipeline_definition
+        self._node = node
+
+    @property
+    def id(self) -> str:
+        return self._node.get('id')
+
+    @property
+    def pipeline(self) -> Dict:
+        return self._pipeline_definition
+
+    @property
+    def dict(self) -> Dict:
+        return self._node
+
+    def get(self, key: str, default_value=None) -> Any:
+        self._validate()
+        return self._node['app_data'].get(key, default_value)
+
+    def set(self, key: str, value: Any):
+        if not key:
+            raise ValueError("Key is required")
+
+        if not value:
+            raise ValueError("Value is required")
+
+        self._node['app_data'][key] = value
+
+    def validate(self) -> list:
+        pass
+
+    def is_valid(self) -> bool:
+        return len(self.validate()) == 0
+
+
+class Pipeline(NodeBase):
+
+    def __init__(self, pipeline_definition: Dict, node: Dict):
+        super().__init__(pipeline_definition, node)
+
+    @property
+    def version(self) -> int:
+        return int(self._node['app_data'].get('version'))
+
+    @property
+    def runtime(self) -> str:
+        return self._node['app_data'].get('runtime') or 'local'
+
+    @property
+    def type(self):
+        type_description_to_type = {'Kubeflow Pipelines': 'kfp',
+                                    'Apache Airflow': 'airflow',
+                                    'Generic': 'generic'}
+
+        pipeline_type_description = self._node['app_data']['properties'].get('runtime')
+        if pipeline_type_description not in type_description_to_type.keys():
+            raise ValueError(f'Unsupported pipeline runtime: {pipeline_type_description}')
+        return type_description_to_type[pipeline_type_description]
+
+    @property
+    def nodes(self) -> list:
+        if 'nodes' not in self._node:
```

Comment:
We should only create these Nodes once if `self._nodes is None` or something, rather than every time the `nodes` property is accessed.